orderBook.js

```diff
@@ -1,5 +1,46 @@
+function findMatch(existingBook, incomingOrder) {
+  return existingBook.filter(bookItem => (bookItem.type !== incomingOrder.type)
+    && !((bookItem.price !== incomingOrder.price) && (bookItem.quantity !== incomingOrder.quantity)))
+}
+
+function filterMatch(existingBook, incomingOrder) {
+  return existingBook.filter(bookItem => (bookItem.type === incomingOrder.type)
+    && (bookItem.quantity <= incomingOrder.quantity)
+    && (bookItem.price === incomingOrder.price))
+}
+
+function priceDecider(existingPrice, incomingPrice) {
+  const priceIndex = existingPrice - incomingPrice
+  if (priceIndex >= 0) {
+    return true
+  }
+}
+
 function reconcileOrder(existingBook, incomingOrder) {
-  throw new Error('Not Implemented Yet')
+  const filteredBook = findMatch(existingBook, incomingOrder)
+
+  if (filteredBook.length && priceDecider(filteredBook[0].price, incomingOrder.price)) {
+
+    let updatedBook = filterMatch(existingBook, incomingOrder)
+
+    const updatedOrder = {
+      type: ((filteredBook[0].quantity > incomingOrder.quantity) ? filteredBook[0].type : incomingOrder.type),
+      quantity: Math.abs(filteredBook[0].quantity - incomingOrder.quantity),
+      price: ((filteredBook[0].price > incomingOrder.price) ? filteredBook[0].price : incomingOrder.price)
+    }
+
+    if (updatedOrder.quantity === 0) {
+      return updatedBook
+    } else {
+      updatedBook = [...updatedBook, updatedOrder]
+      return updatedBook
+    }
+
+  } else {
+    let updatedBook = [...existingBook, incomingOrder]
+    return updatedBook
+
+  }
```

Comment:
The flow here is a little tough to follow, part of that is naming but part of it is the approach. In fact the approach breaks down pretty quickly. If I add another order to one of the tests it goes missing even though it should be preserved and the test would fail.

```JavaScript
    it('fulfills an order and removes the matching order when the book contains a matching order of the same quantity', () => {
      const existingBook = [{ type: 'buy', quantity: 10, price: 6150 }, { type: 'sell', quantity: 10, price: 6000 }]
      const incomingOrder = { type: 'sell', quantity: 10, price: 6150 }

      const updatedBook = reconcileOrder(existingBook, incomingOrder)

      expect(updatedBook).to.deep.equal([{ type: 'sell', quantity: 10, price: 6000 }])
    })
```